models/sudoswap/ethereum/sudoswap_ethereum_events.sql

```diff
@@ -111,70 +132,88 @@ WITH
                     , COALESCE(fu.evt_block_time, pc.block_time) as fee_update_time
                     , pfu.newMultiplier/1e18 as protocolfee --most recent protocolfee, depends on bonding curve to implement it correctly. See explanation in fee table schema.
-                    , pfu.evt_block_time as protocolfee_updatetime
+                    , pfu.evt_block_time as protocolfee_update_time
+                    , pc.protocolfee_recipient
                     , pc.nftcontractaddress
-                    , pc.asset_recip
-                FROM swaps
+                    , coalesce(aru.a, pc.asset_recip) as asset_recip
+                    , coalesce(aru.evt_block_time, pc.block_time) as asset_recip_update_time
+                FROM swaps_with_calldata swaps
                 JOIN pairs_created pc ON pc.pair_address = contract_address --remember swaps from other NFT addresses won't appear!
+                -- we might need to do these joins separately since we're exploding into a lot of rows..
+                -- should not matter a lot since # of changes per pool should be small
                 LEFT JOIN owner_fee_update fu ON swaps.call_block_time >= fu.evt_block_time AND swaps.contract_address = fu.contract_address
                 LEFT JOIN protocol_fee_update pfu ON swaps.call_block_time >= pfu.evt_block_time
+                LEFT JOIN asset_recipient_update aru on swaps.call_block_time >= aru.evt_block_time AND swaps.contract_address = aru.contract_address
             ) a
         ) b
         WHERE ordering = 1 --we want to keep the most recent ownerfee and protocol fee for each individual call (trade)
-    ),
+    )
 
-    swaps_w_traces as (
+    ,swaps_w_traces as (
         -- we traces to get NFT and ETH transfer data because sudoswap doesn't emit any data in events for swaps, so we have to piece it together manually based on trace_address.
         SELECT
             sb.call_block_time
             , sb.call_block_number
             , sb.trade_category
-            , CASE WHEN sb.trade_category = 'Buy' THEN SUM(value)/(1+sb.ownerfee+sb.protocolfee)
-                ELSE SUM(value)/(1-sb.ownerfee-sb.protocolfee)
-                END as base_price
-            , SUM(value) as trade_price --should give total value of the trade (buy or sell)
+            , SUM(
+                CASE WHEN sb.trade_category = 'Buy' -- caller buys, AMM sells
+                THEN (
+                    CASE WHEN tr.from = sb.call_from THEN value -- amount of ETH payed
+                    WHEN (tr.to = sb.call_from AND sb.call_from != sb.asset_recip) THEN -value --refunds unless the caller is also the asset recipient, no way to discriminate there.
+                    ELSE 0 END)
+                ELSE ( -- caller sells, AMM buys
+                    CASE WHEN tr.from = sb.pair_address THEN value -- all ETH leaving the pool, nothing should be coming in on a sell.
+                    ELSE 0 END)
```

Comment:
for sells, where are we adding back in protocol fee and owner fee to get to actual price? since that's subtracted before ETH is transferred out of pool

Reply:
protocol fee is included (cause that also gets transferred out of the pool)
owner fee should not be included cause it's not actually payed.
I'll write my thought framework out in the general comments here. 
Might even be worth including it in a readme in the repo.

Reply:
Hmm, if you look at the bonding curves you'll see that owner fee is added/subtracted from the total price. 

For example here https://github.com/sudoswap/lssvm/blob/main/src/bonding-curves/LinearCurve.sol#L166 

You can `feeMultiplier` of the trade pool is subtracted from the value of ETH actually sent to the seller.

Reply:
The `feeMultiplier` is subtracted from the baseprice to get to the bid price
It's added to the baseprice to get to the ask price.

It's a spread setting instead of fee setting.

I like to think in extremes to clear things up,
If you have a pool with 10 ETH internal spot price, a feemultiplier of 90% and only 1.5 ETH liquidity.
That bid price will be ~1ETH, ask price will be ~19ETH
Someone can come in and sell an NFT to the pool for 1ETH.

The pool gains 1 NFT and pays 1 ETH (I'm ignoring protocol fees for simplicity)
Does it make sense to say the pool **payed and received** 9 ETH in swap fees in this trade?

Reply:
Yes because we want totals? You could make same argument for if protocol fee was 90%, I don't follow